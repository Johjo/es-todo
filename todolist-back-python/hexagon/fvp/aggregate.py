from abc import ABC, abstractmethod
from collections import OrderedDict
from dataclasses import dataclass

from hexagon.shared.type import TaskKey

@dataclass(frozen=True)
class Task:
    key: TaskKey

    def to_do_the_task(self):
        return DoTheTask(key=self.key)

    def to_choose_the_task(self, other_task: 'Task'):
        return ChooseTheTask(main_task_key=self.key, secondary_task_key=other_task.key)


@dataclass(frozen=True)
class NothingToDo:
    pass


@dataclass(frozen=True)
class DoTheTask:
    key: TaskKey


@dataclass(frozen=True)
class ChooseTheTask:
    main_task_key: TaskKey
    secondary_task_key: TaskKey


@dataclass(frozen=True)
class FvpSnapshot:
    task_priorities: OrderedDict[TaskKey, TaskKey]

    def to_primitive_dict(self) -> dict[TaskKey, TaskKey]:
        return {key: value for key, value in self.task_priorities.items()}

    @classmethod
    def from_primitive_dict(cls, data: dict[TaskKey, TaskKey]) -> 'FvpSnapshot':
        return FvpSnapshot(OrderedDict[TaskKey, TaskKey](data))


class FinalVersionPerfectedSession:
    def __init__(self, task_priorities: OrderedDict[TaskKey, TaskKey]):
        self.task_priorities: OrderedDict[TaskKey, TaskKey] = task_priorities

    def which_task(self, open_tasks: list[Task]) -> NothingToDo | DoTheTask | ChooseTheTask:
        ids = [task.key for task in open_tasks]
        tasks = [task for task in open_tasks if self.task_priorities.get(task.key, None) not in ids]
        match tasks:
            case [task]:
                return task.to_do_the_task()
            case [task_1, task_2, *tasks]:
                return task_1.to_choose_the_task(task_2)
            case _:
                return NothingToDo()

    def choose_and_ignore_task(self, id_chosen: TaskKey, id_ignored: TaskKey) -> None:
        self.task_priorities[id_ignored] = id_chosen

    def reset(self) -> None:
        self.task_priorities = OrderedDict()

    def to_snapshot(self) -> FvpSnapshot:
        return FvpSnapshot(self.task_priorities)

    @classmethod
    def from_snapshot(cls, snapshot: FvpSnapshot) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(task_priorities=snapshot.task_priorities)

    @classmethod
    def create(cls) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(OrderedDict())

    def cancel_priority(self, task_key: TaskKey):
        keys = list()
        for (key, value) in self.task_priorities.items():
            if value == task_key:
                keys.append(key)

        for key in keys:
            del self.task_priorities[key]


class FvpSessionSetPort(ABC):
    @abstractmethod
    def save(self, snapshot: FvpSnapshot) -> None:
        pass

    @abstractmethod
    def by(self) -> FvpSnapshot:
        pass
