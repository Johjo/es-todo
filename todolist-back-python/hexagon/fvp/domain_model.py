from collections import OrderedDict
from dataclasses import dataclass


@dataclass
class Task:
    id: int
    name: str

    def to_do_the_task(self):
        return DoTheTask(id=self.id, name=self.name)

    def to_choose_the_task(self, other_task):
        return ChooseTheTask(id_1=self.id, name_1=self.name, id_2=other_task.id, name_2=other_task.name)


@dataclass
class NothingToDo:
    pass


@dataclass
class DoTheTask:
    id: int
    name: str


@dataclass
class ChooseTheTask:
    id_1: int
    name_1: str
    id_2: int
    name_2: str


@dataclass(frozen=True)
class FvpSnapshot:
    task_priorities: OrderedDict[int, int]

    def to_primitive_dict(self) -> dict[int, int]:
        return {key: value for key, value in self.task_priorities.items()}

    @classmethod
    def from_primitive_dict(cls, data: dict[int, int]) -> 'FvpSnapshot':
        return FvpSnapshot(OrderedDict[int, int](data))



class FinalVersionPerfectedSession:
    def __init__(self, task_priorities : OrderedDict[int, int]):
        self.task_priorities : OrderedDict[int, int] = task_priorities

    def which_task(self, open_tasks: list[Task]):
        if not open_tasks:
            return NothingToDo()

        tasks_to_priorize = self._all_tasks_to_priorize(open_tasks)

        if len(tasks_to_priorize) == 1:
            return tasks_to_priorize[0].to_do_the_task()

        return tasks_to_priorize[0].to_choose_the_task(tasks_to_priorize[1])

    def _all_tasks_to_priorize(self, open_tasks : list[Task]) -> list[Task]:
        max_priority = max([self.task_priorities.get(task.id, 0) for task in open_tasks])
        tasks_to_priorize = [task for task in open_tasks if
                             self.task_priorities.get(task.id, max_priority) == max_priority]
        return tasks_to_priorize

    def choose_and_ignore_task(self, id_chosen, id_ignored) -> None:
        if id_chosen not in self.task_priorities:
            self.task_priorities[id_chosen] = self.task_priorities.get(id_ignored, 0) + 1

        self.task_priorities[id_ignored] = self.task_priorities[id_chosen] - 1

    def reset(self) -> None:
        self.task_priorities = OrderedDict()

    def to_snapshot(self) -> FvpSnapshot:
        return FvpSnapshot(self.task_priorities)

    @classmethod
    def from_snapshot(cls, snapshot: FvpSnapshot) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(task_priorities = snapshot.task_priorities)

    @classmethod
    def create(cls) -> 'FinalVersionPerfectedSession':
        return FinalVersionPerfectedSession(OrderedDict())



