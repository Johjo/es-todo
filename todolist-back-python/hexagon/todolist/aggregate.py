from dataclasses import dataclass, replace

from expression import Result, Error, Ok


@dataclass(frozen=True)
class TaskKey:
    value: int


@dataclass
class TaskSnapshot:
    key: TaskKey
    name: str
    is_open: bool


@dataclass
class TodolistSnapshot:
    name: str
    tasks: list[TaskSnapshot]


@dataclass(frozen=True)
class Task:
    key: TaskKey
    name: str
    is_open: bool

    def to_snapshot(self) -> TaskSnapshot:
        return TaskSnapshot(key=self.key, name=self.name, is_open=self.is_open)

    @classmethod
    def from_snapshot(cls, snapshot: TaskSnapshot) -> 'Task':
        return Task(key=snapshot.key, name=snapshot.name, is_open=snapshot.is_open)

    def reword(self, new_name) -> 'Task':
        return replace(self, name=new_name)

    def close_task(self) -> 'Task':
        return replace(self, is_open=False)


@dataclass(frozen=True)
class TodolistAggregate:
    name: str
    tasks: tuple[Task, ...]

    @classmethod
    def create(cls, todolist_name) -> 'TodolistAggregate':
        return TodolistAggregate(name=todolist_name, tasks=())

    @classmethod
    def from_snapshot(cls, snapshot: TodolistSnapshot) -> 'TodolistAggregate':
        return TodolistAggregate(name=snapshot.name, tasks=(*[Task.from_snapshot(task) for task in snapshot.tasks],))

    def to_snapshot(self) -> TodolistSnapshot:
        return TodolistSnapshot(self.name, tasks=[task.to_snapshot() for task in self.tasks])

    def open_task(self, task: Task) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (task,)))

    def close_task(self, key) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")
        return Ok(replace(self, tasks=(*[task.close_task() if task.key == key else task for task in self.tasks],)))

    def reword_task(self, key: TaskKey, new_name: str) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        return Ok(replace(self, tasks=(*[task.reword(new_name) if task.key == key else task for task in self.tasks],)))

    def import_tasks(self, task_snapshots: list[TaskSnapshot]) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (*[Task.from_snapshot(snapshot) for snapshot in task_snapshots],)))

