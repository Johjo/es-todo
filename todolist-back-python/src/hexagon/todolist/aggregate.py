from dataclasses import dataclass, replace

from expression import Result, Error, Ok, Option, Some

from src.hexagon.shared.event import Event, TaskPostponedEvent
from src.hexagon.shared.type import TaskKey, TodolistName, TaskName, TaskExecutionDate, TaskOpen


@dataclass(frozen=True ,eq=True)
class TaskSnapshot:
    key: TaskKey
    name: TaskName
    is_open: TaskOpen
    execution_date: Option[TaskExecutionDate]


@dataclass(frozen=True, eq=True)
class TodolistSnapshot:
    name: TodolistName
    tasks: tuple[TaskSnapshot, ...]


@dataclass(frozen=True, eq=True)
class Task:
    key: TaskKey
    name: TaskName
    is_open: TaskOpen
    execution_date: Option[TaskExecutionDate]

    def to_snapshot(self) -> TaskSnapshot:
        return TaskSnapshot(name=self.name, is_open=self.is_open, key=self.key, execution_date=self.execution_date)

    @classmethod
    def from_snapshot(cls, snapshot: TaskSnapshot) -> 'Task':
        return Task(key=snapshot.key, name=snapshot.name, is_open=snapshot.is_open,
                    execution_date=snapshot.execution_date)

    def reword(self, new_name) -> 'Task':
        return replace(self, name=new_name)

    def close_task(self) -> 'Task':
        return replace(self, is_open=TaskOpen(False))

    def postpone(self, execution_date: TaskExecutionDate):
        return replace(self, execution_date=Some(execution_date))


@dataclass(frozen=True, eq=True)
class TodolistAggregate:
    name: TodolistName
    tasks: tuple[Task, ...]
    events: tuple[Event, ...]

    @classmethod
    def create(cls, todolist_name) -> 'TodolistAggregate':
        return TodolistAggregate(name=todolist_name, tasks=(), events=())

    @classmethod
    def from_snapshot(cls, snapshot: TodolistSnapshot) -> 'TodolistAggregate':
        return TodolistAggregate(name=snapshot.name, tasks=(*[Task.from_snapshot(task) for task in snapshot.tasks],), events=())

    def to_snapshot(self) -> TodolistSnapshot:
        return TodolistSnapshot(self.name, tasks=tuple([task.to_snapshot() for task in self.tasks]))

    def open_task(self, task: Task) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (task,)))

    def close_task(self, key) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")
        return Ok(replace(self, tasks=(*[task.close_task() if task.key == key else task for task in self.tasks],)))

    def reword_task(self, key: TaskKey, new_name: str) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        return Ok(replace(self, tasks=tuple([task.reword(new_name) if task.key == key else task for task in self.tasks],)))

    def import_tasks(self, task_snapshots: list[TaskSnapshot]) -> Result['TodolistAggregate', str]:
        return Ok(replace(self, tasks=self.tasks + (*[Task.from_snapshot(snapshot) for snapshot in task_snapshots],)))

    def postpone_task(self, key: TaskKey, execution_date: TaskExecutionDate) -> Result['TodolistAggregate', str]:
        if not [task for task in self.tasks if task.key == key]:
            return Error(f"The task '{key}' does not exist")

        tasks = tuple([task.postpone(execution_date) if task.key == key else task for task in self.tasks])
        events = self.events + (TaskPostponedEvent(task_key=key, execution_date=execution_date), )
        return Ok(replace(self, tasks=tasks, events=events))

    def uncommited_events(self) -> tuple[Event, ...]:
        return self.events