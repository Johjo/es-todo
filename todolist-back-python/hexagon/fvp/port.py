from abc import ABC, abstractmethod

from hexagon.fvp.domain_model import FvpSnapshot


class FvpSessionSetPort(ABC):
    @abstractmethod
    def save(self, snapshot: FvpSnapshot) -> None:
        pass

    @abstractmethod
    def by(self) -> FvpSnapshot | None:
        pass
