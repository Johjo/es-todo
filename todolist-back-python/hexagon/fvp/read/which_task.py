from abc import ABC, abstractmethod

from hexagon.fvp.domain_model import Task, FinalVersionPerfectedSession, NothingToDo, DoTheTask, ChooseTheTask
from hexagon.fvp.port import FvpSessionRepository


class TaskReader(ABC):
    @abstractmethod
    def all(self) -> list[Task]:
        pass


class WhichTaskQueryContract(ABC):
    @abstractmethod
    def which_task(self) -> NothingToDo | DoTheTask | ChooseTheTask:
        pass


class WhichTaskQuery(WhichTaskQueryContract):
    def __init__(self, set_of_open_tasks: TaskReader, set_of_fvp_sessions: FvpSessionRepository):
        self.set_of_fvp_sessions = set_of_fvp_sessions
        self.set_of_open_tasks = set_of_open_tasks

    def which_task(self):
        session = self._get_or_create_session()
        return session.which_task(self.set_of_open_tasks.all())

    def _get_or_create_session(self):
        snapshot = self.set_of_fvp_sessions.by()
        if snapshot:
            session = FinalVersionPerfectedSession.from_snapshot(snapshot)
        else:
            session = FinalVersionPerfectedSession.create()
        return session
