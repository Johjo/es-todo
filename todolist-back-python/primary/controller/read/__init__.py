from primary.controller.read.count_open_items import count_open_items
from primary.controller.read.export_todo_list_to_markdown import export_todo_list_to_markdown
from primary.controller.read.get_task import get_task
from primary.controller.read.which_task import which_task
from primary.controller.read.current_contexts import current_contexts


