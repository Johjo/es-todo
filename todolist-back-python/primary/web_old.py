import json

from bottle import Bottle, request  # type: ignore

from primary.controller import read

app = Bottle()


@app.route('/rest/todo/<todolist_name>/which_task', method='GET')
def which_task(todolist_name: str) -> str:
    only_inbox: bool = get_string_from_request_get('only_inbox') == '1'
    context: str = get_string_from_request_get('context')
    dependencies = app.dependencies
    response = read.which_task(dependencies, only_inbox=only_inbox, context=context, todolist_name=todolist_name)
    return json.dumps(response)


def get_string_from_request_get(field_name):
    return request.query.getunicode(field_name)
