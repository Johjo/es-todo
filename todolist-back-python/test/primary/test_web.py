import bottle  # type: ignore
import pytest
from approvaltests import verify  # type: ignore
from webtest import TestApp, TestResponse  # type: ignore

from hexagon.fvp.aggregate import ChooseTheTask
from test.primary.controller.read.test_which_task import DependencyListForTest, WhichTaskQueryForTest
from primary.web_old import app as bottle_app


@pytest.fixture
def which_task_query():
    return WhichTaskQueryForTest()


@pytest.fixture
def web_app(which_task_query):
    bottle.debug(True)
    bottle_app._dependencies = DependencyListForTest(which_task_query=which_task_query)
    return bottle_app


@pytest.fixture
def app(web_app):
    return TestApp(web_app)


@pytest.mark.skip(reason="does not work")
def test_rest_which_task(app, which_task_query):
    which_task_query.feed(ChooseTheTask(id_1=1, name_1="buy the milk", id_2=2, name_2="clean the table"))

    try:
        response: TestResponse = app.get('/rest/todo/anytodolist/which_task')
    except Exception as e:
        raise

    assert response.status == '200 OK'
    verify(response.body)
