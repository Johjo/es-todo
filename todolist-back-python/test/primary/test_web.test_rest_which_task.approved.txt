b'{"fvpTasks": [{"id": 1, "name": "buy the milk"}, {"id": 2, "name": "clean the table"}]}'
